### ISSUE TITLE
Doubtful about the authenticity and reproducibility of the results

### ISSUE THREAD

**Issue by JieTian-SALEN** [LiuJMzzZ/OpenNCD#1]
Your work has a very good starting point, and it is indeed a real-world problem that needs to be solved. However, I still have some questions about your code:

1. After running your code, I realized that although the number of new classes is defined as "unknown", you have set a very narrow range of eps for clustering. So although you seem to determine the optimal number of categories by searching for eps, you actually specify the number of known categories by limiting the range of eps. I tried increasing the range of the eps search, and the category count was grossly incorrect (identifying cifar10 as 5 categories) to lead to a completely incorrect result. But with this completely incorrect result, the Unknown ACC is still upwards of 90%, does this indicate a lack of objectivity in your evaluation metrics as well? I think this is very important because, as your paper says, real-world data is a mix of known and unknown categories, and just achieving a metric that looks good by specifying eps on a cifar dataset of known categories doesn't qualify as true real-world research. This is just my personal humble opinion, and I would like to have your further explanation if your setting of eps is something that can be determined in advance before training based on some a priori.

2. after many experiments with the parameters I found that this method relies very much on the initial feature embedding provided by the pre-training weights, and that the number of categories produced by the first clustering is essentially impossible to change during subsequent training. But your omission of the SimCLR pre-training portion gives me reason to wonder if there may have been labels or special sampling methods during the pre-training phase that allowed the model to remember a certain amount of category information that guided the subsequent training process? Of course this is just my unfounded suspicion, and I'd be happy to see a reproducible training process about this part.

**Comment by LiuJMzzZ** [comment]
Thank you for your questions.

1. We checked the code and found that it is an early version in the class estimation part. We have updated the code (main.py) to match the version used in the paper. 
  - The main update is that the min_eps automatically changes by epochs. In the early stages of training, eps is searched in a set with high values (large min_eps). Towards the end of the grouping stage, eps will be searched in a wider range. This contributes to the progressive grouping method, where the group number will decrease until convergence. 
  
  - The second update is to set l_cls (args.reg[1]) to 0 at the grouping stage since it would cause overfitting on the known classes. 
  
- In our method, the estimated number of classes is chosen by the maximum accuracy on the labeled known-class data (MAC-ACC). Recently, some improved methods have been proposed which might be more robust, such as [EMaCS](https://arxiv.org/abs/2212.00334).

2. I agree that representation learning is essential. The first part of our method focuses on learning better representations. However, the pre-trained method plays a minor role in our task. Using k-means on the pre-trained features only achieves approximately 50% clustering accuracy. For consistency, we simply use the same SimCLR pre-training parameter weights as the excellent previous work [ORCA (ICLR'22)](https://github.com/snap-stanford/orca#get-started). Obviously, other unsupervised self-supervised methods may also work.
